pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -60,12 +74,43 @@ def on_trainer_init(
         self.trainer.reload_dataloaders_every_n_epochs = reload_dataloaders_every_n_epochs
         self.trainer._is_data_prepared = False
 
-    def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
-        self.data_fetcher = DataFetcher()
-        self.data_fetcher.setup(train_dataloader)
-        prefetcher_iter = iter(self.data_fetcher)
-        profiled_dl = self.trainer.profiler.profile_iterable(enumerate(prefetcher_iter), "get_train_batch")
-        return profiled_dl
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+        contains_datalaoder_iter = is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True)
+        use_manual_optimization = not self.trainer.lightning_module.automatic_optimization
+        if contains_datalaoder_iter and not use_manual_optimization:
+            raise MisconfigurationException(
+                "When the `training_step` has a `dataloader_idx` argument, you should set "
+                "`automatic_optimization=False` within your LightningModule. "
+            )
+
+        return contains_datalaoder_iter and use_manual_optimization and self.trainer.training
```

Comment:
why's the `self.trainer.training` flag needed?

Reply:
I am following Yifu API. This was activate only for Training. @yifuwang Should I also activate this for other stages ?

Reply:
Hi @tchaton, theoretically other stages can benefit from this as well. But practically, evaluation tends be much less computationally intensive that they can't hide too much communication latency, so a special evaluation step (other than forward) typically isn't justified. I think we can stick to training for now.